File: web/src/engine/playbackPosition.ts
/** Live playback position for the current Track, polled from the player. */
export interface PlaybackPosition {
  currentMillis: number
  totalMillis: number
}

function format(millis: number): string {
  const totalSeconds = Math.max(Math.floor(millis / 1000), 0)
  const minutes = Math.floor(totalSeconds / 60)
  const seconds = totalSeconds % 60
  return `${minutes}:${seconds.toString().padStart(2, '0')}`
}

export function formatCurrent(position: PlaybackPosition): string {
  return format(position.currentMillis)
}

export function formatTotal(position: PlaybackPosition): string {
  return format(position.totalMillis)
}

File: app/src/main/kotlin/com/stopptanz/app/session/PlaybackPosition.kt
package com.stopptanz.app.session

/** Live playback position for the current Track, polled from the player. */
data class PlaybackPosition(val currentMillis: Long, val totalMillis: Long) {
    fun formatCurrent(): String = format(currentMillis)

    fun formatTotal(): String = format(totalMillis)

    private fun format(millis: Long): String {
        val totalSeconds = (millis / 1000).coerceAtLeast(0)
        val minutes = totalSeconds / 60
        val seconds = totalSeconds % 60
        return "%d:%02d".format(minutes, seconds)
    }
}
